server/elasticsearch/src/test/java/org/hawkular/apm/server/elasticsearch/SpanServiceElasticsearchTest.java

```diff
@@ -173,16 +186,210 @@ public void testGetSpanFromEmptyDB() throws StoreException, InterruptedException
+        clientSpanOfServerSpanRoot.setAnnotations(clientAnnotations());
+
+        storeAndWait(null, Arrays.asList(
+                rootClientSpan,
+                serverSpan,
+                serverSpanDescendantRoot,
+                clientSpanOfServerSpanRoot),
+                SpanUniqueIdGenerator::toUnique);
+
+        Trace trace = spanService.getTraceFragment(null, SpanUniqueIdGenerator.getClientId("root"));
+        Assert.assertEquals("root", trace.getId());
+        Assert.assertEquals(1, trace.getNodes().size());
+
+        InteractionNode rootProducerNode = (InteractionNode) trace.getNodes().get(0);
+        Assert.assertThat(rootProducerNode, instanceOf(Producer.class));
+        Assert.assertEquals(1, rootProducerNode.getCorrelationIds().size());
+        Assert.assertEquals("root", rootProducerNode.getCorrelationIds().get(0).getValue());
+        Assert.assertEquals(1, rootProducerNode.getNodes().size());
+        Assert.assertEquals(new HashSet<>(Arrays.asList("root2")),
+                extractCorrelationIds(rootProducerNode.getNodes()));
+
+        InteractionNode descendantConsumerNode = (InteractionNode) rootProducerNode.getNodes().get(0);
+        Assert.assertThat(descendantConsumerNode, instanceOf(Consumer.class));
+        Assert.assertEquals(1, descendantConsumerNode.getCorrelationIds().size());
+        Assert.assertEquals("root2", descendantConsumerNode.getCorrelationIds().get(0).getValue());
+        Assert.assertEquals(0, descendantConsumerNode.getNodes().size());
+    }
+
+    @Test
+    public void testGetTraceFragmentComponent() throws StoreException, InterruptedException {
+        Span rootServerSpan = new Span();
+        rootServerSpan.setId("root");
+        rootServerSpan.setTraceId("root");
+        rootServerSpan.setAnnotations(serverAnnotations());
+
+        Span componentSpan = new Span();
+        componentSpan.setId("component");
+        componentSpan.setParentId("root");
+        componentSpan.setTraceId("root");
+        componentSpan.setName("ejb");
+
+        Span clientComponentSpan = new Span();
+        clientComponentSpan.setId("clientComponent");
+        clientComponentSpan.setParentId("component");
+        clientComponentSpan.setTraceId("root");
+        clientComponentSpan.setAnnotations(clientAnnotations());
+
+        Span clientComponentSpan2 = new Span();
+        clientComponentSpan2.setId("clientComponent2");
+        clientComponentSpan2.setParentId("component");
+        clientComponentSpan2.setTraceId("root");
+        clientComponentSpan2.setAnnotations(clientAnnotations());
+
+        Span descendant = new Span();
+        descendant.setId("descendant");
+        descendant.setParentId("root");
+        descendant.setTraceId("root");
+        descendant.setAnnotations(clientAnnotations());
+
+        Span descendantServer = new Span();
+        descendantServer.setId("descendant");
+        descendantServer.setParentId("descendant");
```

Comment:
This parent id would need to be the same as the 'descendant.getParentId()'.

Just realized that this would cause problems retrieving children from the cache and store - the results shouldn't return server spans, as these would not be children, but do have the same parent id as their client span equivalents. I'll create a jira to investigate this.